sdk/src/azure/core/az_log.c

```diff
@@ -39,39 +43,24 @@ void az_log_set_callback(az_log_message_fn az_log_message_callback)
 // false indicating whether it was logged.
 static bool _az_log_write_engine(bool log_it, az_log_classification classification, az_span message)
```

Comment:
Having this method is good, we don't get any race conditions, but I would simplify to having an inline function returning a callback. So, in _az_log_should_write() you do "return _az_log_get_message_callback() != NULL", and in _az_log_write, you have "_az_log_get_callback()(classification, callback)";

If you want to keep the "engine" function, I would mark it as AZ_INLINE now - previously it was a longer method with iterations, which we didn't neccessarily want to be inlined, and now it is very short. But IMO,
```
AZ_INLINE az_log_message_fn _az_log_get_message_callback()
{
    az_log_message_fn const message_callback = _az_log_message_callback;
    az_log_filter_fn const message_filter_callback = _az_message_filter_callback;

    return _az_log_message_callback != NULL
        && (_az_message_filter_callback != NULL || _az_message_filter_callback(classification))
            ? _az_log_message_callback
            : NULL;
}
```
is better.

Reply:
An example of us doing this is here: https://github.com/Azure/azure-sdk-for-cpp/blob/c5438d623bd95d9bae3256aa1a4c11ceed6c097d/sdk/core/azure-core/src/logging/logging.cpp#L32

Reply:
I am going to defer such changes to a subsequent PR.

Reply:
With an optimizing compiler (-Os) both gcc and clang produce basically identical assembly for both what you have, and what is currently in the PR using `_az_log_write_engine`, when calling `_az_log_write`.

`_az_log_should_write` actually performs slightly better on the original version, compared to yours, when optimized.

But, the current PR performs poorly when not-optimized.